From 7a22e7814662864215b9da0ded11add61599f402 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Florian=20M=C3=BCllner?= <fmuellner@gnome.org>
Date: Fri, 23 Feb 2018 16:56:46 +0100
Subject: [PATCH] Include status-icons in classic session

---
 meson.build | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index 29ac83bb..4b6e3197 100644
--- a/meson.build
+++ b/meson.build
@@ -37,6 +37,7 @@ classic_extensions = [
   'apps-menu',
   'places-menu',
   'launch-new-instance',
+  'status-icons',
   'window-list',
 ]
 
@@ -47,7 +48,6 @@ default_extensions += [
   'heads-up-display',
   'light-style',
   'screenshot-window-sizer',
-  'status-icons',
   'system-monitor',
   'windowsNavigator',
   'workspace-indicator',
-- 
2.51.1

