From 1982ab4218fa3a7ff622fff5af7c15c2e11351f7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Florian=20M=C3=BCllner?= <fmuellner@gnome.org>
Date: Fri, 23 Feb 2018 16:56:46 +0100
Subject: [PATCH] Include top-icons in classic session

---
 meson.build | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index afc0133..78dee5b 100644
--- a/meson.build
+++ b/meson.build
@@ -31,6 +31,7 @@ classic_extensions = [
   'desktop-icons',
   'places-menu',
   'launch-new-instance',
+  'top-icons',
   'window-list'
 ]
 
@@ -49,7 +50,6 @@ all_extensions += [
   'native-window-placement',
   'panel-favorites',
   'systemMonitor',
-  'top-icons',
   'updates-dialog',
   'user-theme'
 ]
-- 
2.32.0

