From b373d600104d6a5c4984521d934c5d02af3a2b99 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Florian=20M=C3=BCllner?= <fmuellner@gnome.org>
Date: Fri, 23 Feb 2018 16:56:46 +0100
Subject: [PATCH] Include status-icons in classic session

---
 meson.build | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index d092830f..14143501 100644
--- a/meson.build
+++ b/meson.build
@@ -35,6 +35,7 @@ classic_extensions = [
   'apps-menu',
   'places-menu',
   'launch-new-instance',
+  'status-icons',
   'window-list',
 ]
 
@@ -45,7 +46,6 @@ default_extensions += [
   'heads-up-display',
   'light-style',
   'screenshot-window-sizer',
-  'status-icons',
   'system-monitor',
   'windowsNavigator',
   'workspace-indicator',
-- 
2.47.1

